### ISSUE TITLE
Idea: Deployment API hooks

### ISSUE THREAD

**Issue by bettse** [newrelic/centurion#110]
Some saas analytics platforms, like New Relic, have an API that can be called during deployment of new versions.  I think it would be cool if centurion had support for something like this.  Maybe just a method that could be overridden to make the API call at start or end of deployment?

**Comment by relistan** [comment]
Generally :+1: but I am concerned about the :boom: that is Capistrano's callback morass.

**Comment by intjonathan** [comment]
Seems like a resque style before_deploy, before_container_stop, before_container_start, after_deploy set would probably get you there.

**Comment by MarkBorcherding** [comment]
I've [started on this](https://github.com/MarkBorcherding/centurion/tree/before-after-callbacks). We need this feature as well to roll nodes in and out of our load balancers.

**Comment by kremso** [comment]
@MarkBorcherding I think that the code might be simpler if you used existing rake API which lets you add before/after callbacks natively, rather than implementing a callback for every point in deployment lifecycle. I think that the only change necessary would be to split the rolling deploy into multiple tasks and passing state between them.

**Comment by MarkBorcherding** [comment]
@kremso I was thinking it would be that easy too when I started, but if you look at [the existing rake tasks](https://github.com/newrelic/centurion/blob/master/lib/tasks/deploy.rake) rolling through the server happens inside one task. You could create a `before_everything` and `after_everything`, but you can't create one around an individual host.

**Comment by kremso** [comment]
That's why I'm suggesting to split the rolling_deploy task into multiple smaller tasks. Something like

````ruby
task :rolling_deploy do
  on_each_docker_host do |server|
    set :target_server, server
    invoke 'deploy:stop'
    invoke 'deploy:start_new'
    invoke 'deploy:health_check'
    invoke 'deploy:cleanup'
  end
end
````

**Comment by MarkBorcherding** [comment]
I see. The problems i see is `invoke`, assuming you're using the `Rake` based `Capistrano`, will only run a task once, and `execute` doesn't run any prereqs...if I remember correctly.

While the end result, would look cleaner, that change is a bit more invasive than I can manage right now.

**Comment by kremso** [comment]
You're right, but you could change the [invoke](https://github.com/newrelic/centurion/blob/master/lib/capistrano_dsl.rb#L87-L89) method to do something like below.

````ruby
Rake::Task[task].invoke(*args)
Rake::Task[task].reenable
Rake::Task[task].all_prerequisite_tasks.each &:reenable
````

**Comment by idleyoungman** [comment]
We're archiving the project, so I'm closing all open issues:
https://github.com/newrelic/centurion/pull/208

Thanks for your contributions.
